diesel/src/macros/mod.rs

```diff
@@ -239,108 +263,235 @@ macro_rules! __diesel_column {
     (
-        $($table_name:ident).+ $(($($pk:ident),+))* {$($body:tt)*}
+        @parse
+        import = [];
+        doc = [$($doc:meta,)*];
+        $($rest:tt)+
     ) => {
         table! {
-            $($table_name).+ $(($($pk),+))* {$($body)*}
-            use $crate::types::*;
+            @parse_body
+            import = [use $crate::types::*;];
+            doc = [$($doc,)*];
+            $($rest)+
+        }
+    };
+    // Add the primary key if it's not present
+    (
+        @parse_body
+        import = [$(use $($import:tt)::+;)+];
+        doc = [$($doc:meta,)*];
+        $($table_name:ident).+ {$($body:tt)*}
+    ) => {
+        table! {
+            @parse_body
+            import = [$(use $($import)::+;)+];
+            doc = [$($doc,)*];
+            $($table_name).+ (id) {$($body)*}
+        }
+    };
+    // Add the schema name if it's not present
+    (
+        @parse_body
+        import = [$(use $($import:tt)::+;)+];
+        doc = [$($doc:meta,)*];
+        $name:ident $(($($pk:ident),+))* {$($body:tt)*}
+    ) => {
+        table! {
+            @parse_body
+            import = [$(use $($import)::+;)+];
+            doc = [$($doc,)*];
+            public . $name $(($($pk),+))* {$($body)*}
         }
     };
-
     // Terminal with single-column pk
     (
+        @parse_body
+        import = [$(use $($import:tt)::+;)+];
+        doc = [$($doc:meta,)*];
         $schema_name:ident . $name:ident ($pk:ident) $body:tt
-        $($imports:tt)+
     ) => {
         table_body! {
-            $schema_name . $name ($pk) $body $($imports)+
+            $schema_name . $name ($pk) $body
+            import = [$(use $($import)::+;)+];
+            doc = [$($doc)*];
         }
     };
 
     // Terminal with composite pk (add a trailing comma)
     (
+        @parse_body
+        import = [$(use $($import:tt)::+;)+];
+        doc = [$($doc:meta,)*];
         $schema_name:ident . $name:ident ($pk:ident, $($composite_pk:ident),+) $body:tt
-        $($imports:tt)+
     ) => {
         table_body! {
-            $schema_name . $name ($pk, $($composite_pk,)+) $body $($imports)+
+            $schema_name . $name ($pk, $($composite_pk,)+) $body
+            import = [$(use $($import)::+;)+];
+            doc = [$($doc)*];
         }
     };
+    // Terminal with invalid syntax
+    (
+        @parse_body
+        import = [$(use $($import:tt)::+;)*];
+        doc = [$($doc:meta,)*];
+        $($rest:tt)*
+    ) => {
+        const ERROR:i32 = "invalid table! syntax";
+    };
+    // Put a parse annotation and empty fields for imports and documentation on top
+    // This is the entry point for parsing the table dsl
+    ($($rest:tt)+) => {
+        table! {
+            @parse
+            import = [];
+            doc = [];
+            $($rest)+
+        }
+    }
 }
 
 #[macro_export]
 #[doc(hidden)]
 macro_rules! table_body {
     (
+        schema_name = $schema_name:ident,
```

Comment:
Can you add comments what these blocks do, like above? (I know there weren't any before but I guess you know what's happening so it'd be awesome to write that down)